# Analysis of Amazon Product Data

After construction of the graph (product-id and link to cross-buys), the following topics will be investigated:
* correlation between in-degrees and salesrank
* and some further smart questions... influence of number of reviews?, are there different sub-graphs like produkt categories?
* 
* 
* 
Data source: https://snap.stanford.edu/data/amazon-meta.html

Data import

In [3]:
# Import graphframes (from Spark-Packages)
from graphframes import *
from pyspark.sql.functions import monotonically_increasing_id, desc


# Get the product data (already pre-processed data in table format)
products = spark.read.csv("/FileStore/tables/ProductData_small.csv", sep=";", header=True)

# The the cross-buy data (similar products)
links = spark.read.csv("/FileStore/tables/LinkData_small.csv", sep=";", header=True)

How many products are we dealing with here?

In [5]:
products.count()

Out[ 5 ]: 10611

quick check of the data formats

In [7]:
display(products.take(3))

Id ASIN Group Salesrank Reviews avg rating Title 21 0790747324 DVD 795 140 4.5 The Time Machine 84 B000063W82 Video 780 14 4.5 The Best of Schoolhouse Rock! - 30th Anniversary Edition 252 B0000262WI Music 581 82 4.5 The Köln Concert

And which and how many groups are there?

In [9]:
display(products.select("Group").distinct())

Group Video Toy DVD Sports Baby Product Video Games Book Music Software

In [10]:
products.select("Group").distinct().count()

Out[ 8 ]: 9

And how many products per category?

In [12]:
display(products.groupBy('Group').count().sort('count', ascending=False))

Group count Video 2933 Music 2855 Book 2535 DVD 2276 Software 5 Toy 4 Sports 1 Baby Product 1 Video Games 1

In [13]:
display(links.take(3))

src dest 0790747324 B00007JMD8 0790747324 6305350221 0790747324 B00004RF9B

The graph will be constructed based on the product data. Each product is identified through its unique *ASIN* (Amazon Standard Identification Number) code

GraphFrame library required *ASIN* to be renamed to *id*, but be careful, there is already an "Id" in the DataFrame

In [15]:
# rename "Id" to "Id-nu" to avoid any confusion (nu = not used)
# rename "ASIN" to "id"
ProductVertices = products.withColumnRenamed("Id", "Id-nu").distinct()
ProductVertices = ProductVertices.withColumnRenamed("ASIN", "id").distinct()
ProductVertices = ProductVertices.withColumnRenamed("avg rating", "avg_rating").distinct()

# check if done correctly
display(ProductVertices.take(3))

Id-nu id Group Salesrank Reviews avg_rating Title 1698 6302796857 Video 424 28 4.5 A Town Like Alice 7524 B00000JWVS DVD 88 58 4.5 Jerry Seinfeld Live on Broadway: I'm Telling You for the Last Time 10025 0393057658 Book 4371 252 4.5 Moneyball: The Art of Winning an Unfair Game

How does the Edge table look like?

In [17]:
display(links.take(1))

src dest 0790747324 B00007JMD8

And how many edges are there?

In [19]:
links.count()

Out[ 13 ]: 52158

Although not always strictly required, it's good practice to have an ID column for the tables. 

Since the product links does not have one yet, create and additional ID:

In [21]:
# 2. Create Edges
links = links.withColumn("link-id", monotonically_increasing_id())

display(links.take(3))

src dest link-id 0790747324 B00007JMD8 0 0790747324 6305350221 1 0790747324 B00004RF9B 2

Salesrank and number of reviews are read as string. Change type to int for calculations.

In [23]:
# change type of features "Salesrank" and "Reviews" to int
ProductVertices = ProductVertices.withColumn('Salesrank', ProductVertices.Salesrank.cast('int'))
ProductVertices = ProductVertices.withColumn('Reviews', ProductVertices.Reviews.cast('int'))
ProductVertices = ProductVertices.withColumn('avg rating', ProductVertices.avg_rating.cast('float'))
display(ProductVertices.take(1))

Id-nu id Group Salesrank Reviews avg_rating Title avg rating 1698 6302796857 Video 424 28 4.5 A Town Like Alice 4.5

Rename links to similar products ar *src* and *dst*

In [25]:
# rename destination "dest" to "dst"
links = links.withColumnRenamed('dest', 'dst')


Now comes the moment of truth! Let's see if the graph can be constructed!

In [27]:
# Build the GraphFrame based on products and links

ProductGraph = GraphFrame(ProductVertices, links)

# print a sample of the vertices and edges
print("Products: %d" % ProductGraph.vertices.count())
print
print("SAMPLE PRODUCT DATA:")
print(ProductGraph.vertices.take(1))
print

print("Links: %d" % ProductGraph.edges.count())
print
print("SAMPLE LINK DATA:")
print(ProductGraph.edges.take(1))
print

Products: 10611
SAMPLE PRODUCT DATA:
[Row(Id-nu='1698', id='6302796857', Group='Video', Salesrank=424, Reviews=28, avg_rating='4.5', Title='A Town Like Alice', avg rating=4.5)]
Links: 52158
SAMPLE LINK DATA:
[Row(src='0790747324', dst='B00007JMD8', link-id=0)]
 Out[ 17 ]: <function print>

The Graph is constructed. Do the fancy analyses!

1) assess the number of in-degress per product and compare it to the salesrank

Plot the next chunk in the following way:

you can look at the distribution of the inDegree-values by clicking on the icon to the right of 'Raw table' and then chosing a histogram, values = 'inDegree' and  30 bins

We can see that most of the inDegree values are between 0 and 2

In [31]:
display(ProductGraph.inDegrees.sort('inDegree', ascending=False))
#inDegRank = ProductGraph.inDegrees.sort('inDegree', ascending=False)

id inDegree B00008LDNZ 78 B00005JMAH 55 B00023P4I8 53 B00003CXCT 52 B000634DCW 45 B000068DBC 37 6304765266 36 B00005ATZT 34 B00001QEE7 33 B00005JMIJ 33 6305368171 33 1594480001 33 B0000TB01Y 32 B00003CXA2 30 630522577X 29 B0001NBLVI 29 B000059HB7 29 B0002CR04A 28 B0001I561E 28 6305696071 27 0684801523 27 B00005JMJG 27 B00003CX95 27 B000022TSH 27 B0002IVN9W 26 B000286RKW 26 B000000XB9 26 0316769487 24 B00005JMQZ 24 B00005JN4W 23 B0000DJZ8R 23 B000065DWB 23 B00005QFDX 23 B00003CXCQ 22 B0000026NF 22 B00000F14B 22 1400032717 22 B00012QM8G 22 B00005JKGO 21 B00005Y1XY 21 B00005V3Z4 21 B0002JP2OI 21 B0001NBMH6 21 B00001QEE6 21 B00005LKHZ 20 B00005LC4A 20 B00004RF97 20 B00005KARK 20 B00000K0DS 20 0790731479 20 B00005YUPQ 20 0446310786 20 B00008975Z 20 0694003611 19 B00001MXXH 19 B00005JMQW 19 B00008CMR4 19 B00005NTNW 19 0790729628 19 B0000AOV4I 19 B000067J1P 19 B00003CXE7 19 0140177396 19 6305280770 19 B0002YLCOM 19 B0002MPQRS 18 B00004R9A0 18 0805047905 18 B00003CX8Y 18 0399226907 18 B00005JKOI 18 B00005Q79A 18 0792839749 18 B0002I84JO 18 0800141660 18 6305168857 18 B00049QQJQ 18 0399501487 18 B00006FDCP 18 B00006RCO1 18 0066620996 18 B00000K31Q 18 B0002VEPRQ 17 0449214923 17 B00003JRCQ 17 B0001Z4OXS 17 B00003CXBU 17 0792165055 17 B0001L3MI0 17 B00000JS62 17 B00005NQJJ 17 6305144168 17 B00005JKDR 16 B00004XPPG 16 B00005PJ8V 16 0790729377 16 B00004KD24 16 B000055Y0X 16 B000002ADT 16 B0001932ZU 16 0064401847 16 0440237688 16 B00004U104 16 B00003CXC7 16 6305216088 16 B00019072G 16 B000053VBK 16 B000055ZF6 16 B000023VTP 16 0783225725 16 B00000DHZJ 15 1568322267 15 1568328478 15 0783117523 15 B00003CXD3 15 6305212112 15 6304696566 15 B00002CGGH 15 0440498058 15 B0007TKNL0 15 B00074DY0M 15 B000035Z1Z 15 0792833287 15 B00005S6K8 15 B00004VVNB 15 6303144535 15 B00003CXAT 15 0553210092 15 B00005R87R 14 B00005Q4CS 14 B00004RF9B 14 B00023P4II 14 B000056WQX 14 B00003CX9E 14 B00006JMRE 14 B00004TDTO 14 B0001XAPZ6 14 B00005ASGF 14 6305428476 14 B0000A98ZP 14 B00005JM02 14 B00005YUPN 14 0790749815 14 B00000ILEC 14 B0000058HV 14 B00005N5S3 14 B0000TG9E2 14 B00005AAA9 14 B00005QJEF 14 1568328605 14 B00007J88T 14 B00007LVCM 14 B00005B6JZ 14 630529142X 14 B00005JMUA 14 B00006AL1E 14 0064430227 14 B0000BWVA3 14 6303467814 14 0451526341 14 B0000062XS 14 B00009W0WM 14 0060254920 14 B00008ZZ7B 14 B00000G3AZ 13 B0000648WV 13 0071381392 13 6304711891 13 B000087F7D 13 B0000542C9 13 0783240171 13 B000065UFD 13 B00005O3VC 13 B00007E2F5 13 B0002WZTOI 13 6304907729 13 B000002KAR 13 B00005PJ8S 13 0783227884 13 B000002G3Y 13 B00005MP58 13 0312422156 13 079284615X 13 B00005LZT9 13 B00001QEE9 13 B000055XM8 13 B00005MLVA 13 B00007GZR1 13 B00007JMD8 13 015602943X 13 B000000W5P 13 B00005JKHN 13 B0006H32DY 13 B00006IR69 13 0689835604 13 B0000033WI 13 6304711913 13 B00001QEE8 13 6305428042 13 B0000AQS6R 13 B00009YXET 13 0738920762 13 0375810455 13 0609610570 13 B00005JN0Q 13 B0000047CX 13 B00005221M 13 B00005JSHW 13 B00008NRI8 13 0780621611 13 B00005JN2Z 13 078322592X 13 B00000I21X 13 B00004Y62W 13 B000002TTP 13 B00003RQNJ 12 0060256656 12 B0000YEEGM 12 0440414806 12 B0000CDL6V 12 0767830555 12 0738920770 12 6304698526 12 0790740060 12 B00004TJKK 12 081297106X 12 0800141741 12 B000067FP3 12 B000001AMZ 12 B00008438T 12 6304696612 12 B00004R99Q 12 B0000BZK1R 12 B00019330O 12 B00006JSS8 12 630460291X 12 B00000K3CQ 12 B00005PJ8O 12 B0000TLA9G 12 B00004RF9D 12 B000065V3P 12 B00005JKFR 12 B00009NHBM 12 B00003CXDC 12 B0006Z2NAY 12 0345342968 12 0767803434 12 6304698690 12 0783233493 12 B00003CXDD 12 B00003CWL6 12 B000002UAO 12 B000094J7Z 12 B0000A02TZ 12 B00000IBME 12 B00005LOKQ 12 B000065U1M 12 B000063K1U 12 B000002KHJ 12 B0000541AK 12 B00003CXCZ 12 B00005N7YZ 12 B00002RAPT 12 0790731037 12 189176702X 12 B00005JKG1 12 B0002I832C 12 B0000648WU 12 630227608X 12 0394800168 12 B00000IQW3 12 6305511055 11 B0001HAGSI 11 B000002U8G 11 1568328443 11 B00004WG2F 11 6304696531 11 6305161917 11 B00005QFE1 11 B00009NHC0 11 B00005B234 11 B000002OO0 11 B00006DEF9 11 B00003CXC5 11 B00005J6UP 1

In [32]:
maxNumInDeg = ProductGraph.inDegrees.groupBy().max('inDegree')
display(maxNumInDeg)

maxNumInDegValue = str(maxNumInDeg.collect()[0]['max(inDegree)'])
#print(maxNumInDegValue)

maxInDegreeProduct = ProductGraph.inDegrees.filter('inDegree = {}'.format(maxNumInDegValue))
display(maxInDegreeProduct)


id inDegree B00008LDNZ 78

In [33]:
# join back to get product info based on max inDegree
ProductMaxInDegree = ProductGraph.vertices.join(maxInDegreeProduct, 'id')
display(ProductMaxInDegree)



id Id-nu Group Salesrank Reviews avg_rating Title avg rating inDegree B00008LDNZ 548091 DVD 110 83 5.0 Laura 5.0 78

Plot the following chunk in this way:
Plot this as a bar chart with 'Plot Options':
 - Keys: InDegree
 - Values: Salesrank
 
Unfortunately you can't chose to have a logarithmic y-axis, but you do see that at least for low in-degree values the sales rank is rather bad.

Alternative: scatterplot inDegree vs. salesrank and grouping by product group (Keys=Group)
Update: Boxplot avg_rating vs. Salesrank is can now be created, but does not provide much info

In [35]:
# repeat exercise from above but with entire list
selection = ProductGraph.inDegrees.sort('inDegree', ascending=False)
#display(selection)

ProductSortInDeg = ProductGraph.vertices.join(selection, 'id').sort('inDegree', ascending=False)
display(ProductSortInDeg)

id Id-nu Group Salesrank Reviews avg_rating Title avg rating inDegree B00008LDNZ 548091 DVD 110 83 5.0 Laura 5.0 78 B000068DBC 546689 DVD 109 633 4.5 Pulp Fiction (Collector's Edition) 4.5 37 6304765266 97579 DVD 351 152 4.5 While You Were Sleeping 4.5 36 B00005ATZT 547040 DVD 1020 140 4.5 The Fugitive (Special Edition) 4.5 34 B00001QEE7 104775 DVD 49 160 4.5 The Little Mermaid (Limited Issue) 4.5 33 6305368171 276371 DVD 236 500 4.0 You've Got Mail 4.0 33 B00003CXA2 547272 DVD 185 530 4.0 Forrest Gump 4.0 30 630522577X 548182 DVD 253 180 4.5 My Fair Lady 4.5 29 B000059HB7 4388 DVD 1877 79 4.5 Rio Bravo 4.5 29 B000022TSH 544390 DVD 1108 154 4.5 Chinatown 4.5 27 6305696071 545472 DVD 540 193 4.0 Pretty Woman (10th Anniversary Edition) 4.0 27 B00003CX95 166635 DVD 233 745 4.5 Braveheart 4.5 27 0684801523 199628 Book 956 934 4.0 The Great Gatsby 4.0 27 B000000XB9 327081 Music 87 163 5.0 """Creedence Clearwater Revival - Chronicle, Vol. 1: The 20 Greatest Hits""" 5.0 26 0316769487 98756 Book 60 2568 4.0 The Catcher in the Rye 4.0 24 B000065DWB 46130 Music 417 31 4.0 Foreigner - Complete Greatest Hits 4.0 23 B00005QFDX 220569 DVD 330 115 4.5 Sesame Street - The Best of Elmo 4.5 23 B00000F14B 216263 DVD 351 123 4.5 The Music Man (Special Edition) 4.5 22 B0000026NF 3499 Music 153 180 4.5 Journey - Greatest Hits 4.5 22 B00003CXCQ 402685 DVD 679 247 4.5 Snow White and the Seven Dwarfs (Disney Special Platinum Edition) 4.5 22 B00001QEE6 444195 DVD 904 68 4.5 Lady and the Tramp (Limited Issue) 4.5 21 B00005V3Z4 546221 DVD 176 848 4.5 Donnie Darko 4.5 21 B00005Y1XY 429734 Music 333 47 4.5 Cars - Complete Greatest Hits 4.5 21 B00008975Z 270909 DVD 695 407 4.5 Reservoir Dogs 4.5 20 B00005YUPQ 28804 DVD 389 129 4.5 Baby Einstein - Baby Van Gogh - World of Colors 4.5 20 B00004RF97 96590 DVD 792 104 4.5 The Philadelphia Story 4.5 20 B00005KARK 453361 DVD 353 140 4.5 Dumbo (60th Anniversary Edition) 4.5 20 B00005LKHZ 164067 DVD 343 236 4.5 The NeverEnding Story 4.5 20 B00005LC4A 545385 DVD 1260 158 4.5 Charade 4.5 20 0446310786 537519 Book 111 1414 4.5 To Kill a Mockingbird 4.5 20 0790731479 138550 DVD 633 126 4.5 Beetlejuice 4.5 20 B00005NTNW 548266 DVD 768 78 4.5 The Outlaw Josey Wales 4.5 19 B00003CXE7 520382 DVD 1702 1513 4.5 Gladiator 4.5 19 B00008CMR4 547333 DVD 1798 38 4.0 Big Jake 4.0 19 B00001MXXH 548318 DVD 652 295 4.5 Ferris Bueller's Day Off 4.5 19 0140177396 341570 Book 126 1000 4.5 Of Mice and Men (Penguin Great Books of the 20th Century) 4.5 19 0694003611 502086 Book 156 339 4.5 Goodnight Moon (Board Book) 4.5 19 0790729628 545235 DVD 392 812 4.0 Blade Runner (The Director's Cut) 4.0 19 6305280770 546760 DVD 1915 79 4.5 The King and I 4.5 19 B000067J1P 545271 DVD 1331 357 4.5 The Sound of Music (Single Disc Full Screen Edition) 4.5 19 B00006RCO1 544551 DVD 979 162 5.0 All About Eve (Special Edition) 5.0 18 B00005Q79A 191370 DVD 1169 166 4.5 What's Eating Gilbert Grape 4.5 18 0800141660 546277 DVD 443 250 4.5 Sense and Sensibility 4.5 18 0066620996 154855 Book 29 363 4.5 Good to Great: Why Some Companies Make the Leap... and Others Don't 4.5 18 0792839749 12389 DVD 384 125 3.5 Chitty Chitty Bang Bang (Full Screen Edition) 3.5 18 B00000K31Q 18824 DVD 1379 387 4.5 10 Things I Hate About You 4.5 18 0805047905 502784 Book 171 172 5.0 Brown Bear, Brown Bear, What Do You See? 5.0 18 B00003CX8Y 191667 DVD 439 354 4.5 Beauty and the Beast (Disney Special Platinum Edition) 4.5 18 0399226907 180888 Book 279 164 4.5 The Very Hungry Caterpillar board book 4.5 18 0399501487 35512 Book 143 1101 4.0 Lord of the Flies 4.0 18 B00005JKOI 137583 DVD 2582 75 4.0 Pretty in Pink 4.0 18 6305168857 152642 DVD 154 234 4.5 Young Frankenstein (Special Edition) 4.5 18 B00006FDCP 59459 DVD 1582 224 4.5 One Flew Over the Cuckoo's Nest (Two-Disc Special Edition) 4.5 18 0449214923 482263 Book 215 260 4.5 Think and Grow Rich 4.5 17 6305144168 544508 DVD 972 276 4.5 Dead Poets Society 4.5 17 B00003JRCQ 424377 DVD 462 204 4.5 Persuasion 4.5 17 B00000JS62 547782 DVD

Plot inDegree vs. Salesrank for Books only
* plot as scatter plot with LOESS smoother

In [37]:
# try to differentiate by product group
ProductSortInDeg = ProductGraph.vertices.join(selection, 'id').sort('inDegree', ascending=False).filter("Group = 'Book'")
display(ProductSortInDeg)

id Id-nu Group Salesrank Reviews avg_rating Title avg rating inDegree 0684801523 199628 Book 956 934 4.0 The Great Gatsby 4.0 27 0316769487 98756 Book 60 2568 4.0 The Catcher in the Rye 4.0 24 0446310786 537519 Book 111 1414 4.5 To Kill a Mockingbird 4.5 20 0694003611 502086 Book 156 339 4.5 Goodnight Moon (Board Book) 4.5 19 0140177396 341570 Book 126 1000 4.5 Of Mice and Men (Penguin Great Books of the 20th Century) 4.5 19 0066620996 154855 Book 29 363 4.5 Good to Great: Why Some Companies Make the Leap... and Others Don't 4.5 18 0805047905 502784 Book 171 172 5.0 Brown Bear, Brown Bear, What Do You See? 5.0 18 0399501487 35512 Book 143 1101 4.0 Lord of the Flies 4.0 18 0399226907 180888 Book 279 164 4.5 The Very Hungry Caterpillar board book 4.5 18 0449214923 482263 Book 215 260 4.5 Think and Grow Rich 4.5 17 0440237688 188765 Book 428 2617 4.5 The Giver 4.5 16 0064401847 42948 Book 1258 655 4.5 Bridge to Terabithia 4.5 16 0440498058 291216 Book 414 872 4.5 A Wrinkle in Time 4.5 15 0553210092 317053 Book 368 348 3.5 The Scarlet Letter 3.5 15 0451526341 53175 Book 275 990 4.5 Animal Farm 4.5 14 0064430227 206514 Book 1032 76 5.0 Harold and the Purple Crayon 50th Anniversary Edition (Purple Crayon Books) 5.0 14 0060254920 268369 Book 148 211 5.0 Where the Wild Things Are 5.0 14 0689835604 335281 Book 250 90 4.5 Where Is Baby's Belly Button? 4.5 13 0609610570 429788 Book 143 157 3.5 Execution: The Discipline of Getting Things Done 3.5 13 0375810455 220181 Book 209 20 4.5 Too Big for Diapers (Too Big Board Books) 4.5 13 0312422156 425735 Book 133 517 4.5 Middlesex: A Novel 4.5 13 0071381392 18173 Book 239 355 4.5 The No-Cry Sleep Solution: Gentle Ways to Help Your Baby Sleep Through the Night 4.5 13 0060256656 239578 Book 360 394 4.0 The Giving Tree 4.0 12 0440414806 528497 Book 761 2922 4.5 Holes (Yearling Newbery) 4.5 12 189176702X 532007 Book 1725 8 4.5 "How to be the Leader of the Pack...And have Your Dog Love You For It. (""How to"" booklets from Dog's Best Friend)" 4.5 12 0394800168 261029 Book 895 148 4.5 Green Eggs and Ham (I Can Read It All by Myself Beginner Books) 4.5 12 0345342968 29543 Book 81 1069 4.0 Fahrenheit 451 4.0 12 0439064864 216201 Book 445 2422 5.0 Harry Potter and the Chamber of Secrets (Book 2) 5.0 11 043935806X 429348 Book 110 5545 4.5 Harry Potter and the Order of the Phoenix (Book 5) 4.5 11 0446677450 162283 Book 37 1637 4.0 Rich Dad, Poor Dad: What the Rich Teach Their Kids About Money--That the Poor and Middle Class Do Not! 4.0 10 0787960756 33690 Book 89 61 4.5 The Five Dysfunctions of a Team: A Leadership Fable 4.5 10 0486280861 296016 Book 122 13 5.0 Easy Spanish Phrase Book: Over 770 Basic Phrases for Everyday Use 5.0 10 1573441449 436250 Book 467 32 4.5 The Ultimate Guide to Cunnilingus: How to Go Down on a Woman and Give Her Exquisite Pleasure (Ultimate Everything!!!) 4.5 10 0345446798 240297 Book 2251 77 4.5 The Other End of the Leash 4.5 10 0440413281 40399 Book 930 350 4.5 Bud, Not Buddy (Newbery Medal Winner, 2000) 4.5 10 076360013X 33247 Book 295 209 4.5 Guess How Much I Love You (Guess How Much I Love You) 4.5 10 0440136482 532042 Book 161 351 3.5 Holy Blood, Holy Grail 3.5 10 0142001740 337971 Book 26 996 4.0 The Secret Life of Bees 4.0 10 0671723650 53885 Book 64 377 4.5 How to Win Friends & Influence People 4.5 10 0486291138 515301 Book 290 35 4.5 1001 Most Useful Spanish Words (Beginners' Guides) 4.5 10 0684852861 357720 Book 215 189 4.5 First, Break All the Rules: What the World's Greatest Managers Do Differently 4.5 9 0440439884 105186 Book 1457 400 4.5 Island of the Blue Dolphins 4.5 9 0439136350 404634 Book 356 2396 5.0 Harry Potter and the Prisoner of Azkaban (Book 3) 5.0 9 0142000663 211438 Book 201 517 4.5 The Grapes of Wrath : (Centennial Edition) 4.5 9 0689711816 179199 Book 1414 251 4.5 From the Mixed-Up Files of Mrs. Basil E. Frankweiler 4.5 9 0446677469 234402 Book 384 155 3.5 Rich Dad's Guide to Investing: What the Rich Invest in, That the Poor and the Middle Class Do 

Plot the following chunk in this way:
Plot this as a bar chart with 'Plot Options':
 - Keys: OutDegree
 - Values: Salesrank

In [39]:
# repeat exercise from above for outDegree
selection = ProductGraph.outDegrees.sort('outDegree', ascending=True)
#display(selection)

ProductSortOutDeg = ProductGraph.vertices.join(selection, 'id').sort('OutDegree', ascending=True)
display(ProductSortOutDeg)

id Id-nu Group Salesrank Reviews avg_rating Title avg rating outDegree 6303031803 481374 Video 1778 3 3.5 A Gnome Named Gnorm 3.5 1 6300215350 394817 Video 2142 19 4.5 Water 4.5 1 6300218228 481265 Video 721 4 5.0 Woman Called Golda 5.0 1 6303471617 241928 Video 4095 5 4.0 Rolling Thunder 4.0 1 B0000714F7 475706 Video 447 7 5.0 The Forsyte Saga - First Generation 5.0 1 B00000F5ZW 282259 Video 1014 26 4.5 Scavenger Hunt 4.5 1 6300270246 424280 Video 4661 6 3.0 Deal of the Century 3.0 1 B00005MM6X 57827 Video 4135 0 0.0 First Aid Pet Emergency: Dogs 0.0 1 6301973135 162750 Video 1011 8 5.0 Of Human Bondage 5.0 1 6300271382 328876 Video 3023 7 4.0 Pete Kelly's Blues 4.0 1 6301883055 522541 Video 1991 4 5.0 Voices of Sarafina 5.0 1 630180256X 443818 Video 869 27 5.0 Dogs in Space 5.0 1 B00003CM1K 255215 Video 4534 9 4.0 Navy SEALs Workout 4.0 1 6305046425 452509 Video 2568 2 1.5 Huckleberry Finn 1.5 1 189379203X 301720 Video 1215 8 3.5 Embrace the Moon Tai Chi Chuan 3.5 1 1889679127 125359 Video 3288 2 4.0 Weight Loss with Kundalini Yoga 4.0 1 6303018009 69997 Video 4047 3 3.5 Stop the World - I Want to Get Off 3.5 1 0792844629 14249 Video 3810 10 3.5 The Importance of Being Earnest 3.5 1 B00005BGPK 209410 Video 4063 4 4.5 BMX Bandits 4.5 1 1573309788 165886 Video 3596 12 4.5 The Real World You Never Saw - Boston and Seattle 4.5 1 6301972228 464116 Video 4898 4 4.5 The Next Voice You Hear 4.5 1 6302816599 151078 Video 4020 16 4.5 Goodbye, My Lady 4.5 1 1930612001 80821 Video 2051 5 4.0 Transformations 4.0 1 0800121090 293341 Video 2536 9 3.0 Calendar Girl 3.0 1 1573625035 281510 Video 4730 8 3.0 Morgan Stewart's Coming Home 3.0 1 6304771746 313402 Video 4555 2 4.0 The Far Pavilions (EP Version) 4.0 1 6304458835 233314 Video 4896 9 5.0 Hotel de Love 5.0 1 B00005BJY7 91975 Video 3508 15 5.0 Lil Bow Wow: Beware of Video 5.0 1 B00000EZTT 199712 Video 2228 7 4.5 Greased Lightning 4.5 1 6303451144 90972 Video 1494 16 4.0 The Legend of the Lone Ranger 4.0 1 6302816661 292947 Video 619 29 4.5 Twice Upon a Time 4.5 1 1578071305 54593 Video 1468 2 5.0 NOVA: Lost at Sea - The Search for Longitude 5.0 1 6302277760 64681 Video 202 7 4.5 Pure Luck 4.5 1 0966444310 74029 Video 2027 4 4.5 Positive Postnatal Fitness - A Complete Post Pregnancy Workout 4.5 1 0780613252 395036 Video 905 10 5.0 In the Wild - Dolphins With Robin Williams 5.0 1 6305198152 16835 Video 4804 12 4.5 Project Grizzly 4.5 1 0736033386 456306 Video 915 10 4.5 Beth Shaw: YogaFit Workout 4.5 1 B00004XPJ9 433551 Video 3489 0 0.0 South African Safari: From Shamwari to the Wine Country 0.0 1 6302427355 82364 Video 4661 6 4.5 Pink Floyd: La Carrera Panamericana 4.5 1 6303910327 281355 Video 81 3 5.0 Homophobia in the Workplace 5.0 1 6302275997 35205 Video 2833 10 4.5 Rover Dangerfield 4.5 1 B000056AWK 263628 Video 2858 4 2.5 The Starlost - Beginning 2.5 2 6301967372 71302 Video 4195 12 4.5 Cold Turkey 4.5 2 6302650402 321627 Video 1966 8 5.0 The Mary Tyler Moore Show: Chuckles Bites The Dust 5.0 2 093040758X 8748 Video 3053 0 0.0 RUMI: THE WINGS OF LOVE 0.0 2 6305074070 465375 Video 1127 20 4.0 Song of Norway 4.0 2 6302424879 231746 Video 3209 8 3.5 Young Winston 3.5 2 6303344267 452637 Video 1216 20 4.5 Denis Leary: No Cure for Cancer 4.5 2 B00005A1VR 492822 Video 1521 6 4.5 Beverly Hills Family Robinson 4.5 2 0767832604 186252 Video 1354 3 4.5 Party of Five: Richer, Poorer, Sickness, and Health 4.5 2 6304425880 8422 Video 4022 3 5.0 Water Dog 5.0 2 630456810X 196637 Video 4259 1 5.0 Drywall Hanging & Taping 5.0 2 B00005A1V9 492861 Video 1000 7 4.0 Pearl (The Miniseries) 4.0 2 6305806489 408677 Video 4300 3 3.5 Ultimate Massage 3.5 2 B00004WG9F 77339 Video 2288 7 5.0 One Foot in the Grave - Who Will Buy 5.0 2 6303477151 247370 Video 4107 4 4.5 Kingfish: A Story of Huey P. Long 4.5 2 630176627X 168785 Video 2028 17 4.5 China Beach 4.5 2 6302804752 114560 Video 4329 12 3.5 Atlantis, the Lost Continent 3.5 2 B00004TJO8 378000 Video 1464 7 2.0 Bugs Bunny's Halloween Hijinks 2.0 2 B0000

outDegree is rather boring, since most of the products get assigned 5 products by default that are presumingly similar, as you can see here:

In [41]:
display(ProductSortOutDeg.groupBy("outDegree").count())

outDegree count 1 41 2 43 3 48 4 43 5 10343

plot the output of the next chunk in the following way:
chose a scatter plot with options:
  - values: Salesrank, Reviews
  - LOESS bandwidth 0.47
  
and apply to all rows.
We can see a very weak dependency, suggesting that the less reviews a product has, the less likely it is to be a top selling product (a potential causal orientation might actually be inverse here...)

In [43]:
# relation of number of reviews and sales rank:
display(ProductGraph.vertices.select("Reviews", "Salesrank").sort("Reviews", ascending = False))


Reviews Salesrank 5545 110 5539 420 5539 2135 5039 170 5039 1533 5034 746 4924 404 4924 661 4924 327 3839 3221 3520 3494 3520 1616 3518 1676 3518 187 3051 3468 3049 19 2925 487 2925 2264 2922 761 2617 428 2568 3735 2568 60 2562 2872 2562 1599 2562 511 2562 28 2422 3807 2422 445 2419 1227 2396 2569 2396 356 2295 142 2280 551 2167 313 2097 85 2097 565 2091 17 2072 1388 2026 3649 2026 662 1985 214 1983 648 1936 2199 1927 654 1927 1647 1894 2165 1869 1073 1867 4574 1855 1983 1826 4207 1826 337 1826 280 1826 1576 1818 420 1814 1425 1814 4748 1814 4836 1811 267 1811 757 1726 2985 1697 2059 1695 1378 1695 160 1695 1233 1688 3162 1637 3459 1637 37 1635 471 1627 455 1552 31 1551 898 1540 1790 1513 4615 1513 770 1513 1702 1501 4273 1460 486 1458 676 1432 1817 1423 401 1414 111 1414 3089 1414 3396 1410 4220 1382 3159 1377 3085 1377 42 1318 709 1315 46 1300 493 1300 4000 1293 1567 1291 599 1285 2218 1279 126 1278 539 1278 1038 1278 3505 1261 1662 1244 782 1223 133 1223 2679 1200 851 1199 123 1196 969 1191 975 1186 1667 1185 4107 1182 3634 1173 85 1173 4231 1169 1491 1169 3014 1156 723 1156 131 1143 1587 1143 4979 1142 339 1108 238 1103 4761 1101 4372 1101 1592 1101 143 1101 4831 1100 1302 1094 1934 1094 1372 1094 2592 1091 571 1069 81 1069 3681 1060 712 1059 622 1051 2714 1048 1734 1046 3853 1030 1991 1019 3708 1016 1234 1015 1596 1011 428 1006 1578 1004 1179 1002 1842 1000 126 996 26 996 2121 994 4287 993 722 990 275 990 227 990 762 989 1301 988 4679 982 491 974 539 974 1401 972 69 969 489 961 1018 958 1320 957 33 955 3786 955 862 952 894 934 956 933 4429 931 427 927 273 926 1403 926 2206 925 298 925 184 924 661 921 3346 918 3334 903 2326 881 105 880 836 873 1476 872 414 869 128 869 712 863 3956 861 293 861 4226 855 2986 855 151 855 419 849 156 848 176 847 754 840 1321 837 2471 835 782 829 1012 827 174 826 919 824 1097 822 526 819 723 817 3538 817 2136 816 32 814 579 812 392 812 4690 810 2959 810 533 809 1075 808 232 805 424 805 562 803 3057 803 42 801 1704 799 1395 796 102 794 395 793 804 792 796 792 1929 786 3503 786 47 785 176 785 3981 784 1963 781 242 780 1592 780 3147 777 668 772 4751 771 182 771 1303 771 342 770 3429 770 3280 765 3019 761 1305 760 1790 758 2502 757 3049 755 1240 755 998 755 4122 753 767 751 114 746 21 746 1324 746 619 745 3092 745 305 745 233 744 497 734 1072 732 195 730 3870 730 552 729 669 728 135 722 4453 721 2471 720 1388 720 1605 718 1215 712 970 712 492 712 75 709 996 709 2025 707 61 707 1356 705 194 703 1825 702 3734 702 1606 701 265 701 1572 697 2744 697 1874 693 1836 692 2816 692 295 692 1414 688 429 688 1716 685 4396 683 89 682 2955 682 107 681 4579 681 157 680 1060 679 3367 677 269 677 4492 677 965 677 224 676 185 676 4493 676 3331 676 2857 675 3417 675 4968 674 628 670 125 670 853 655 1258 653 1750 652 3147 650 2877 648 2403 647 1572 647 2917 647 223 647 1344 642 596 639 2697 639 2295 636 111 636 2743 636 482 634 1179 633 713 633 1023 633 109 631 134 631 3319 630 1875 630 581 628 1992 628 2494 626 4113 626 412 624 4030 622 2488 620 4213 619 661 619 2627 617 2937 617 4550 616 517 612 575 612 1973 612 4262 612 1433 611 1906 610 3334 610 3049 609 1229 609 80 609 2903 609 377 608 454 606 1154 603 1348 602 3300 602 4061 601 3237 600 4832 600 174 600 333 598 4275 598 1032 597 1739 597 2825 595 660 595 3111 594 1524 593 585 593 4272 592 1466 591 1944 590 2505 589 1229 587 67 586 4191 586 2621 586 1042 586 38 584 2217 583 405 583 3961 583 2191 583 4930 582 1049 582 3884 582 84 579 217 578 1280 578 986 576 2171 574 3807 572 483 572 273 572 1043 571 685 571 2162 569 708 567 4820 567 570 567 113 567 1599 566 1126 565 2452 564 470 563 4410 562 1532 562 268 562 3320 562 529 561 2457 561 4694 559 539 557 255 557 300 556 786 556 3715 555 2986 555 432 554 1100 553 2345 552 3727 551 2171 549 4023 548 695 548 178 546 1252 546 413 545 2439 544 4229 544 1962 543 2405 543 1683 543 764 541 117 541 515 539 1165 538 2299 537 988 537 855 537 234 535 2942 534 1073 531 681 531 3160 530 185 530 4769 529 149 529 1636 528 2979 527 1

what is the correlation coefficient between the two?

In [45]:
ProductGraph.vertices.stat.corr("Reviews", "Salesrank")

Out[ 36 ]: -0.18539669378148713

trying to find products which have an edge to another product which in turn has an edge to the first product:

In [47]:
ProductsWithBalancedSimilarity = ProductGraph.find("(v1)-[e1]->(v2); (v2)-[e2]->(v1)")
display(ProductsWithBalancedSimilarity)

v1 e1 v2 e2 List(21, 0790747324, DVD, 795, 140, 4.5, The Time Machine, 4.5) List(0790747324, B00007JMD8, 0) List(114681, B00007JMD8, DVD, 1076, 104, 4.5, Journey to the Center of the Earth, 4.5) List(B00007JMD8, 0790747324, 11138) List(84, B000063W82, Video, 780, 14, 4.5, The Best of Schoolhouse Rock! - 30th Anniversary Edition, 4.5) List(B000063W82, 156949407X, 5) List(239294, 156949407X, Video, 1223, 24, 5.0, Schoolhouse Rock! - Grammar Rock, 5.0) List(156949407X, B000063W82, 22808) List(296, 0385504209, Book, 19, 3049, 3.5, The Da Vinci Code, 3.5) List(0385504209, 0671027360, 20) List(376858, 0671027360, Book, 31, 1552, 4.0, Angels & Demons, 4.0) List(0671027360, 0385504209, 34818) List(296, 0385504209, Book, 19, 3049, 3.5, The Da Vinci Code, 3.5) List(0385504209, 0671027387, 22) List(424705, 0671027387, Book, 84, 429, 4.0, Deception Point, 4.0) List(0671027387, 0385504209, 38171) List(448, 0312966091, Book, 607, 141, 4.5, Three To Get Deadly : A Stephanie Plum Novel (A Stephanie Plum Novel), 4.5) List(0312966091, 0312990456, 40) List(471848, 0312990456, Book, 304, 414, 4.5, One for the Money (A Stephanie Plum Novel), 4.5) List(0312990456, 0312966091, 41794) List(448, 0312966091, Book, 607, 141, 4.5, Three To Get Deadly : A Stephanie Plum Novel (A Stephanie Plum Novel), 4.5) List(0312966091, 0312971346, 41) List(104366, 0312971346, Book, 1401, 264, 4.5, High Five (A Stephanie Plum Novel), 4.5) List(0312971346, 0312966091, 10118) List(448, 0312966091, Book, 607, 141, 4.5, Three To Get Deadly : A Stephanie Plum Novel (A Stephanie Plum Novel), 4.5) List(0312966091, 0312976275, 42) List(475036, 0312976275, Book, 787, 324, 4.5, Hot Six : A Stephanie Plum Novel (A Stephanie Plum Novel), 4.5) List(0312976275, 0312966091, 41970) List(448, 0312966091, Book, 607, 141, 4.5, Three To Get Deadly : A Stephanie Plum Novel (A Stephanie Plum Novel), 4.5) List(0312966091, 0312980140, 43) List(158327, 0312980140, Book, 1084, 274, 4.0, Seven Up (A Stephanie Plum Novel), 4.0) List(0312980140, 0312966091, 15295) List(457, B0000296JB, Music, 2439, 545, 4.5, Make Yourself, 4.5) List(B0000296JB, B00005QG9J, 45) List(119794, B00005QG9J, Music, 2488, 622, 4.5, Morning View, 4.5) List(B00005QG9J, B0000296JB, 11691) List(480, B0000296J9, Music, 1310, 42, 5.0, Gunfighter Ballads & Trail Songs, 5.0) List(B0000296J9, B0000026AG, 50) List(219235, B0000026AG, Music, 2015, 26, 5.0, Johnny Horton - Greatest Hits, 5.0) List(B0000026AG, B0000296J9, 21158) List(480, B0000296J9, Music, 1310, 42, 5.0, Gunfighter Ballads & Trail Songs, 5.0) List(B0000296J9, B0000024YO, 51) List(438745, B0000024YO, Music, 2293, 4, 4.5, Marty Robbins - All-Time Greatest Hits, 4.5) List(B0000024YO, B0000296J9, 39218) List(486, B0000296J3, Music, 1060, 46, 4.5, Stardust, 4.5) List(B0000296J3, B00004U2G7, 55) List(320882, B00004U2G7, Music, 709, 49, 5.0, Red Headed Stranger, 5.0) List(B00004U2G7, B0000296J3, 30023) List(486, B0000296J3, Music, 1060, 46, 4.5, Stardust, 4.5) List(B0000296J3, B00008BXK3, 57) List(394855, B00008BXK3, Music, 496, 15, 4.5, The Essential Willie Nelson, 4.5) List(B00008BXK3, B0000296J3, 36016) List(577, 0451163966, Book, 758, 316, 4.5, One Flew Over the Cuckoo's Nest, 4.5) List(0451163966, 0393312836, 68) List(196284, 0393312836, Book, 1073, 534, 4.5, A Clockwork Orange (Norton Paperback Fiction), 4.5) List(0393312836, 0451163966, 19148) List(600, B000002V0T, Music, 3430, 55, 4.5, American Patriot, 4.5) List(B000002V0T, B00005QJDD, 72) List(162660, B00005QJDD, Music, 3422, 59, 4.0, God Bless America, 4.0) List(B00005QJDD, B000002V0T, 15643) List(622, B000002V07, Music, 1659, 79, 4.5, So Tonight That I Might See, 4.5) List(B000002V07, B000002TMY, 75) List(107745, B000002TMY, Music, 2588, 24, 4.5, Among My Swan, 4.5) List(B000002TMY, B000002V07, 10454) List(622, B000002V07, Music, 1659, 79, 4.5, So Tonight That I Might See, 4.5) List(B000002V07, B000003TA7, 78) List(269022, B000003TA7, Music, 4920, 55, 4.5, Blind, 4.5) List(B000003TA7, B000002V07, 25516) List(656

how many product-pairs reference each other? can we just calculate it like this?

In [49]:
ProductsWithBalancedSimilarity.count()

Out[ 32 ]: 11202

does this mean, we have 11202 product-pairs that reference each other? Let's test if our data frame of mutually referencing products lists the pairs once or twice by looking at an example:

In [51]:
display(ProductsWithBalancedSimilarity.filter("(v1.id = '0790747324' and v2.id = 'B00007JMD8') or (v2.id = '0790747324' and v1.id = 'B00007JMD8')"))

v1 e1 v2 e2 List(21, 0790747324, DVD, 795, 140, 4.5, The Time Machine, 4.5) List(0790747324, B00007JMD8, 0) List(114681, B00007JMD8, DVD, 1076, 104, 4.5, Journey to the Center of the Earth, 4.5) List(B00007JMD8, 0790747324, 11138) List(114681, B00007JMD8, DVD, 1076, 104, 4.5, Journey to the Center of the Earth, 4.5) List(B00007JMD8, 0790747324, 11138) List(21, 0790747324, DVD, 795, 140, 4.5, The Time Machine, 4.5) List(0790747324, B00007JMD8, 0)

So the pairs are listed twice, which kind of makes sense, seeing how we constructed the data frame. This means there are 11202/2 = 5601 product-pairs that reference each other.

Now we want to explore if the "strongly connected components" of our graph are evenly distributed across all groups, which would mean that the probability of a DVD referencing to another DVD is the same as it referencing to e.g. a book, or if the groups tend to stick to themselves.

Here we first calculate the strongly connected components
(small warning: this command takes quite a while, ~1.17 hours, >250 jobs):

In [54]:
connectedComp = ProductGraph.stronglyConnectedComponents(maxIter=3)
#connectedComp.select("id", "component").groupBy("component").count().sort(desc("count")).show()

#ProductGraph.connectedComponents()
display(connectedComp)

Id-nu id Group Salesrank Reviews avg_rating Title avg rating component 194466 6305973385 DVD 3145 108 3.5 Paula Abdul - Cardio Dance 3.5 26 45292 B000002NJH Music 295 282 4.5 Paint the Sky with Stars: The Best of Enya 4.5 29 265694 B00000K2VU Music 2504 31 4.5 Cheap Thrills (Exp) 4.5 34359738398 100395 B00005Q6OS Music 1254 43 4.5 Those Who Tell the Truth Shall Die, Those Who Tell the Truth Shall Live Forever 4.5 77309411361 526151 055357342X Book 853 670 4.5 A Storm of Swords (A Song of Ice and Fire, Book 3) 4.5 137438953476 477878 0316769509 Book 4230 137 4.5 Nine Stories 4.5 146028888066 301867 6303418899 Video 3006 5 4.5 There Goes a Motorcycle 4.5 146028888086 169155 6304400551 Video 3815 216 4.5 Mary Poppins 4.5 206158430228 190765 156686903X Book 3495 102 4.0 Final Fantasy VIII Official Strategy Guide 4.0 171798691859 355037 B00004U38Q Music 2436 11 5.0 The Very Best of Perry Como 5.0 94489280561 3499 B0000026NF Music 153 180 4.5 Journey - Greatest Hits 4.5 23 463953 0156711427 Book 2866 84 4.0 A Passage to India 4.0 377957122049 426240 B00006F7S4 Music 831 450 4.5 Sea Change 4.5 395136991288 411914 6304973063 Video 912 2 5.0 Return to Hockeytown - Detroit Red Wings 1998 Stanley Cup Champions 5.0 335007449114 389335 B000088USB Video 1292 249 4.5 Kiki's Delivery Service 4.5 429496729640 211661 B000000XGX Music 3608 13 5.0 The Best of the Staple Singers [Stax] 5.0 455266533408 88272 0817440593 Book 2423 53 5.0 John Shaw's Nature Photography Field Guide 5.0 463856467978 270055 B00000JH89 Music 1749 443 4.5 Running With Scissors 4.5 532575944733 309546 B00000656E Video 4919 7 3.5 Compromising Positions 3.5 592705486881 450459 B00003CXAY DVD 764 99 4.5 Koyaanisqatsi / Powaqqatsi (2 Pack) 4.5 249108103200 134195 B000002HCO Music 3151 128 4.5 Debut 4.5 661424963617 72153 6301966708 Video 1383 8 4.5 Duet for One 4.5 695784701974 167890 B000000ORS Music 3382 108 4.5 Miss Saigon (Original 1989 London Cast) 4.5 23 515087 0156007754 Book 2899 251 4.0 Blindness (Harvest Book) 4.0 790273982468 106195 1568327447 Video 3195 1 5.0 The Magic School Bus: Ready, Set, Dough 5.0 206158430226 185264 B0000640SB DVD 2509 206 4.0 Men in Black (Deluxe Edition) 4.0 23 177261 B00004TJS2 DVD 4472 91 4.5 Five Deadly Venoms 4.5 678604832796 201515 B000002GC6 Music 2927 21 5.0 Nothing to Fear 5.0 919123001375 173166 0316341517 Book 295 60 4.0 Mythology 4.0 395136991241 225671 B00005NRNG DVD 718 170 4.5 Pilates Workout for Dummies 4.5 8589934634 18173 0071381392 Book 239 355 4.5 The No-Cry Sleep Solution: Gentle Ways to Help Your Baby Sleep Through the Night 4.5 51539607567 210663 B00000J7AQ Music 4661 15 5.0 Honky Tonk Heroes 5.0 214748364856 534204 B00005OSOA Video 3170 1 4.0 Sacred Ground 4.0 1030792151083 277079 0553351397 Book 1271 53 5.0 Peace Is Every Step : The Path of Mindfulness in Everyday Life 5.0 1039382085640 36726 B00004R99R Video 4459 134 4.0 Hercules 4.0 1125281431585 436706 0735203199 Book 2645 87 4.0 How to Get Into the Top MBA Programs 4.0 1228360646667 24187 B00008AOX3 DVD 1666 188 4.5 Dawson's Creek - The Complete First Season 4.5 1271310319672 633 0689714920 Book 2745 54 4.5 Misty of Chincoteague 4.5 1348619730949 124556 0790743442 Video 2323 46 5.0 Yankee Doodle Dandy 5.0 1348619730952 365652 B000002UAI Music 1915 185 4.0 Beatles for Sale 4.0 1417339207716 182969 B00003L9CJ DVD 955 75 5.0 Mr. Smith Goes to Washington 5.0 23 266708 0316178314 Book 4008 27 4.0 Six Thinking Hats 4.0 1477468749826 118519 0310201063 Book 2736 102 3.5 The Purpose-Driven Church 3.5 1511828488195 10914 0790719029 Video 1676 35 4.0 Free Willy 4.0 1554778161163 264458 0783115202 DVD 409 79 4.0 Three Amigos! 4.0 23 478727 0553372114 Book 4135 36 4.5 A Path with Heart : A Guide Through the Perils and Promises of Spiritual Life 4.5 1614907703302 127076 B000056VOO DVD 3569 50 4.5 Mystery Science Theater 3000 - The Wild World of Batwoman 4.5 687194767391 241859 6302712947 Video 2056 0 0.0 The Lawrence Welk Show - Tribute to Glenn Miller & His Orchestra 0.0 1391

The connected components are marked by the same value in the column "component". So if we group the data frame by "component" and "Group" and use the function "count()", we will have as many rows per "component" value as the number of different groups belonging to this cluster.

In [56]:
countGroupComponent= connectedComp.groupBy("component", "Group").count().sort("component")
display(countGroupComponent)

component Group count 0 Book 5 1 Book 1 2 Book 1 3 Book 1 4 Book 1 5 Book 1 6 Book 1 7 Book 8 8 Book 2 9 Video 9 11 Video 1 12 Book 7 13 Book 11 14 Book 1 15 Video 1 16 Video 1 17 Video 1 18 Video 2 19 Video 16 20 Video 1 21 Video 1 22 Video 2 23 Book 8 23 DVD 679 23 Video 18 23 Music 283 24 DVD 1 25 Video 1 26 DVD 1 26 Video 4 27 Music 4 28 Music 2 29 Music 6 30 Music 1 31 Music 1 32 Video 1 33 DVD 1 33 Video 2 34 Video 1 35 Music 1 36 DVD 1 37 Music 2 38 Music 1 39 Music 2 40 Video 5 41 DVD 1 42 DVD 1 43 Video 1 43 Music 4 44 Video 1 45 DVD 4 46 Music 4 47 DVD 10 48 Music 1 49 Music 4 50 DVD 9 51 DVD 3 51 Video 2 52 Music 1 8589934592 Book 1 8589934593 Book 2 8589934594 Book 1 8589934595 Book 4 8589934596 Video 1 8589934597 Video 1 8589934598 DVD 1 8589934599 DVD 1 8589934600 Video 1 8589934601 Video 1 8589934602 Book 1 8589934603 Video 2 8589934604 Video 1 8589934605 Video 1 8589934606 Book 1 8589934607 Video 3 8589934608 Video 1 8589934609 Video 1 8589934610 Video 1 8589934611 Video 1 8589934612 Video 1 8589934613 Video 2 8589934614 Video 1 8589934617 Music 1 8589934617 DVD 7 8589934619 Music 1 8589934620 Music 3 8589934621 Music 1 8589934622 Music 5 8589934623 Music 3 8589934624 Music 1 8589934626 DVD 1 8589934628 Music 1 8589934630 DVD 4 8589934631 Music 2 8589934632 Video 1 8589934633 Music 2 8589934634 Book 1 8589934634 Video 15 8589934634 DVD 5 8589934635 Music 4 8589934636 Music 4 8589934637 Music 1 8589934638 DVD 1 8589934639 Video 1 17179869184 Book 1 17179869185 Book 8 17179869186 Book 1 17179869187 Book 6 17179869188 Book 6 17179869188 DVD 1 17179869189 Book 1 17179869190 Book 1 17179869191 DVD 4 17179869192 DVD 1 17179869193 Book 1 17179869194 Book 3 17179869195 Video 1 17179869196 Book 3 17179869197 Video 1 17179869198 Book 1 17179869199 Book 12 17179869200 Video 1 17179869202 Video 3 17179869203 Video 2 17179869204 Video 1 17179869205 Video 1 17179869206 Video 1 17179869207 Video 1 17179869208 Video 1 17179869209 DVD 1 17179869210 DVD 5 17179869211 Music 1 17179869212 Music 1 17179869213 Music 1 17179869214 Music 1 17179869216 Music 2 17179869217 Music 5 17179869218 Music 2 17179869219 Music 1 17179869220 Music 6 17179869221 Music 2 17179869222 Music 1 17179869223 Music 2 17179869224 Music 6 17179869225 Music 11 17179869226 Video 1 17179869227 DVD 1 17179869228 Video 1 17179869229 Music 1 17179869230 DVD 2 17179869231 DVD 8 17179869232 Music 4 17179869233 DVD 1 17179869234 Music 2 17179869235 Music 1 17179869237 Music 2 17179869238 DVD 1 17179869238 Book 1 17179869239 Music 4 17179869240 Music 1 17179869241 Music 4 17179869243 Music 4 25769803776 Book 3 25769803777 Book 1 25769803778 Book 1 25769803779 Book 2 25769803780 Book 1 25769803781 Book 4 25769803782 Book 1 25769803783 Book 3 25769803784 Book 3 25769803785 Book 1 25769803786 DVD 2 25769803787 Video 1 25769803788 Video 1 25769803789 Music 1 25769803790 Video 1 25769803791 Video 1 25769803792 Video 1 25769803793 Video 2 25769803794 Video 1 25769803795 Video 2 25769803796 Video 3 25769803797 DVD 1 25769803798 Video 1 25769803799 Music 1 25769803800 Music 10 25769803801 Music 1 25769803803 Music 3 25769803804 Video 1 25769803805 Video 1 25769803806 Music 1 25769803807 DVD 1 25769803808 DVD 3 25769803809 Music 1 25769803810 Music 6 25769803811 DVD 1 25769803812 Video 1 25769803813 Video 1 25769803814 Video 1 25769803816 Video 1 25769803817 DVD 2 25769803818 Video 1 25769803819 Video 1 25769803820 Video 5 25769803821 Music 1 25769803822 Music 2 25769803825 Music 3 25769803826 DVD 2 25769803827 DVD 1 34359738368 Book 1 34359738369 Book 5 34359738370 Book 1 34359738371 Book 3 34359738372 Book 1 34359738373 DVD 1 34359738373 Book 8 34359738374 Book 6 34359738375 Book 3 34359738376 Book 1 34359738377 Book 9 34359738378 Video 1 34359738379 Book 1 34359738380 Book 1 34359738381 Book 1 34359738382 Book 7 34359738383 Video 1 34359738384 Video 1 34359738385 Video 1 34359738386 Video 1 34359738387 Video 1 34359738388 Video 1 34359738389 Video 1 34359738390 Video 1 343

Now how many groups are taking part in each cluster?

In [58]:
countComponentsNumberOfGroups = countGroupComponent.groupBy("component").count().sort(desc("count"))
display(countComponentsNumberOfGroups)

component count 23 4 1357209665545 3 8589934634 3 730144440322 3 180388626482 3 300647710759 3 266287972394 3 360777252871 2 283467841564 2 1468878815251 2 206158430229 2 137438953478 2 120259084313 2 584115552308 2 206158430222 2 1503238553636 2 635655159829 2 146028888077 2 33 2 120259084315 2 188978561035 2 463856468026 2 807453851683 2 197568495668 2 17179869188 2 26 2 566935683117 2 240518168599 2 103079215150 2 42949672988 2 94489280515 2 403726925860 2 43 2 51539607595 2 17179869238 2 103079215159 2 154618822666 2 403726925871 2 463856467996 2 163208757271 2 738734374927 2 309237645341 2 223338299392 2 223338299435 2 249108103184 2 816043786248 2 566935683114 2 68719476754 2 1391569403943 2 103079215122 2 841813590031 2 128849018893 2 85899345925 2 85899345968 2 171798691898 2 549755813926 2 481036337168 2 704374636575 2 51 2 266287972352 2 51539607614 2 34359738373 2 120259084322 2 103079215156 2 395136991247 2 335007449126 2 541165879344 2 661424963604 2 8589934617 2 60129542190 2 163208757277 2 343597383724 2 1537598292018 2 386547056678 2 111669149742 2 1486058684465 2 601295421501 2 876173328445 2 137438953488 2 137438953514 2 420906795031 2 77309411352 2 77309411362 2 180388626459 2 51539607587 1 781684047881 1 687194767365 1 781684047884 1 687194767368 1 687194767369 1 687194767370 1 652835029000 1 687194767374 1 687194767375 1 687194767376 1 687194767379 1 687194767380 1 687194767381 1 687194767383 1 51539607588 1 687194767386 1 687194767387 1 687194767389 1 687194767390 1 687194767391 1 687194767394 1 687194767396 1 687194767397 1 687194767398 1 687194767399 1 687194767400 1 687194767401 1 687194767403 1 695784701953 1 781684047885 1 695784701958 1 481036337198 1 481036337199 1 481036337200 1 489626271745 1 489626271746 1 489626271749 1 489626271754 1 489626271756 1 489626271757 1 489626271758 1 489626271759 1 489626271760 1 489626271761 1 489626271762 1 489626271764 1 489626271765 1 489626271768 1 489626271771 1 489626271772 1 489626271773 1 489626271775 1 489626271776 1 489626271777 1 489626271786 1 489626271787 1 489626271790 1 489626271793 1 489626271795 1 781684047886 1 498216206342 1 498216206346 1 137438953485 1 137438953486 1 137438953489 1 137438953491 1 137438953487 1 137438953497 1 137438953490 1 137438953501 1 137438953492 1 137438953493 1 137438953494 1 137438953511 1 137438953498 1 137438953499 1 137438953500 1 146028888066 1 137438953502 1 137438953503 1 137438953504 1 137438953505 1 137438953506 1 137438953508 1 137438953509 1 420906795020 1 137438953515 1 137438953516 1 137438953517 1 137438953519 1 137438953522 1 137438953523 1 146028888065 1 420906795045 1 146028888068 1 644245094408 1 644245094409 1 644245094410 1 644245094411 1 644245094412 1 644245094413 1 644245094414 1 644245094415 1 644245094417 1 644245094419 1 644245094420 1 644245094421 1 644245094422 1 644245094425 1 644245094426 1 644245094427 1 644245094428 1 644245094431 1 644245094432 1 644245094433 1 644245094438 1 644245094439 1 644245094440 1 644245094441 1 51539607590 1 652835029002 1 532575944718 1 532575944719 1 532575944720 1 532575944721 1 532575944722 1 532575944723 1 532575944724 1 532575944725 1 532575944726 1 532575944728 1 532575944729 1 532575944731 1 532575944732 1 532575944733 1 532575944734 1 532575944736 1 532575944737 1 532575944739 1 532575944740 1 532575944744 1 532575944745 1 532575944746 1 532575944747 1 532575944748 1 532575944749 1 532575944750 1 532575944751 1 541165879297 1 541165879298 1 541165879299 1 541165879304 1 541165879305 1 541165879306 1 541165879307 1 541165879308 1 412316860453 1 412316860457 1 412316860456 1 420906795011 1 420906795008 1 420906795009 1 420906795010 1 420906795022 1 420906795012 1 420906795015 1 420906795016 1 420906795017 1 420906795018 1 420906795019 1 1254130450482 1 420906795046 1 420906795023 1 420906795025 1 420906795026 1 420906795027 1 420906795028 1 420906795030 1 420906795032 1 420906795033 1 420906795034 1 420906795035 1 420906795037 1 420906795039 1 420906795040

Next we get an overview how many clusters there are with members of *n* groups:

In [60]:
numGroupsCount = countComponentsNumberOfGroups.withColumnRenamed("count", "NumberOfGroupsForComponent").groupBy("NumberOfGroupsForComponent").count()
display(numGroupsCount)

NumberOfGroupsForComponent count 4 1 3 6 2 77 1 6667

So how many clusters are there in total? Just adding up the values in column "count" in the last chunk results in the number of clusters:

In [62]:
display(numGroupsCount.agg({"count": "sum"}))

sum(count) 6751

6751 is a very high number, seeing that there are just 10611 products in our dataset in total.

And what is the size distribution of the clusters in general (independent which groups its members belong to)?

In [65]:
countConnected = connectedComp.groupBy("component").count().sort(desc("count"))
display(countConnected)

component count 23 988 51539607567 29 60129542167 25 8589934634 21 154618822687 21 197568495616 19 163208757252 19 94489280561 19 19 16 343597383724 15 51539607594 14 51539607602 14 214748364856 14 94489280564 13 85899345942 13 163208757279 13 77309411362 12 17179869199 12 103079215153 11 154618822666 11 13 11 17179869225 11 206158430226 10 47 10 120259084303 10 403726925830 10 137438953501 10 180388626432 10 25769803800 10 34359738373 9 34359738377 9 103079215122 9 274877906976 9 9 9 188978561051 9 94489280515 9 120259084320 9 197568495621 9 137438953475 9 50 9 42949672967 9 240518168599 9 369367187467 9 7 8 386547056654 8 60129542190 8 266287972385 8 51539607582 8 223338299441 8 8589934617 8 171798691841 8 94489280554 8 137438953474 8 17179869231 8 68719476740 8 17179869185 8 68719476739 8 111669149708 8 77309411337 8 51539607615 8 446676598811 8 455266533417 8 94489280555 8 386547056678 8 51539607564 7 927712935969 7 240518168582 7 51539607614 7 309237645335 7 17179869188 7 541165879333 7 558345748483 7 128849018896 7 77309411328 7 223338299403 7 481036337154 7 240518168622 7 188978561046 7 214748364802 7 584115552283 7 68719476754 7 103079215111 7 146028888081 7 206158430208 7 77309411350 7 34359738382 7 12 7 395136991239 7 360777252871 6 317827579904 6 549755813926 6 94489280559 6 472446402563 6 798863917093 6 429496729606 6 171798691857 6 120259084290 6 17179869224 6 94489280512 6 51539607560 6 249108103172 6 481036337184 6 42949672974 6 300647710752 6 94489280553 6 60129542145 6 420906795028 6 223338299442 6 206158430222 6 652835029007 6 137438953493 6 25769803810 6 120259084322 6 188978561026 6 463856468026 6 120259084301 6 489626271746 6 51539607563 6 17179869187 6 111669149696 6 652835029006 6 910533066761 6 463856468010 6 850403524615 6 85899345960 6 34359738403 6 566935683081 6 395136991234 6 618475290648 6 463856467997 6 343597383711 6 790273982502 6 420906795026 6 29 6 111669149742 6 128849018886 6 283467841594 6 34359738374 6 60129542149 6 85899345968 6 51539607597 6 111669149740 6 730144440322 6 206158430209 6 17179869220 6 420906795045 5 266287972357 5 532575944751 5 618475290629 5 120259084297 5 42949672963 5 343597383685 5 395136991288 5 257698037768 5 85899345939 5 111669149738 5 231928234003 5 420906795019 5 489626271764 5 481036337168 5 25769803820 5 257698037781 5 283467841562 5 438086664196 5 94489280520 5 40 5 51 5 249108103193 5 85899345964 5 180388626482 5 335007449107 5 326417514535 5 111669149737 5 60129542179 5 128849018928 5 352187318307 5 472446402599 5 171798691891 5 8589934622 5 790273982485 5 429496729621 5 463856467975 5 128849018918 5 68719476743 5 283467841588 5 180388626459 5 17179869210 5 154618822703 5 283467841564 5 103079215150 5 335007449098 5 420906795049 5 17179869217 5 34359738369 5 60129542178 5 249108103200 5 120259084332 5 755914244123 5 180388626446 5 249108103198 5 360777252904 5 43 5 1365799600132 5 60129542163 5 515396075564 5 197568495624 5 438086664222 5 403726925860 5 26 5 0 5 352187318322 5 68719476736 5 111669149746 5 343597383726 5 712964571151 5 292057776169 5 128849018890 5 120259084300 5 369367187473 4 42949672987 4 274877906985 4 678604832775 4 17179869241 4 850403524654 4 575525617700 4 137438953498 4 137438953476 4 46 4 1005022347272 4 377957122062 4 34359738399 4 455266533412 4 94489280514 4 1211180777479 4 661424963598 4 51539607570 4 51539607583 4 1022202216497 4 257698037775 4 197568495643 4 8589934630 4 438086664251 4 240518168605 4 103079215155 4 395136991260 4 111669149741 4 8589934595 4 25769803781 4 111669149751 4 532575944714 4 103079215148 4 223338299418 4 17179869239 4 8589934636 4 403726925851 4 266287972394 4 103079215159 4 128849018885 4 103079215105 4 103079215126 4 68719476745 4 103079215137 4 197568495633 4 317827579944 4 42949673003 4 171798691855 4 901943132161 4 678604832798 4 532575944736 4 111669149763 4 249108103205 4 27 4 128849018919 4 94489280541 4 996432412673 4 644245094441 4 953482739729 4 171798691872 4 111669149719 4 85899345965 4

So there is only one really huge cluster of strongly connected components with almost 1000 members. The next biggest one has only a size of 29.

How many of the 6751 clusters have more than 3 members?

In [68]:
countConnected.filter("count > 3").count()

Out[ 40 ]: 352